lbrynet/core/Session.py

```diff
@@ -98,112 +93,109 @@ def __init__(self, blob_data_payment_rate, db_dir=None,
         self.upnp_redirects = []
-
         self.wallet = wallet
-        self.dht_node_class = dht_node_class
-        self.dht_node = None
-
+        self.dht_node_class = dht_node_class or node.Node
+        self.dht_node = self.dht_node_class(
+            udpPort=self.dht_node_port,
+            node_id=self.node_id,
+            externalIP=self.external_ip
+        )
+        self.peer_finder = peer_finder or DHTPeerFinder(self.dht_node, self.peer_manager)
+        self.hash_announcer = hash_announcer or DHTHashAnnouncer(self.dht_node, self.peer_port)
+        self.blob_manager = blob_manager or DiskBlobManager(self.hash_announcer,
+                                                    self.blob_dir,
+                                                    self.db_dir)
+        self.blob_tracker = self.blob_tracker_class(self.blob_manager,
+                                                    self.peer_finder,
+                                                    self.dht_node)
         self.base_payment_rate_manager = BasePaymentRateManager(blob_data_payment_rate)
-        self.payment_rate_manager = None
+
         self.payment_rate_manager_class = payment_rate_manager_class or NegotiatedPaymentRateManager
         self.is_generous = is_generous
+        self.payment_rate_manager = self.payment_rate_manager_class(
+            self.base_payment_rate_manager,
+            self.blob_tracker,
+            self.is_generous)
 
+    @defer.inlineCallbacks
     def setup(self):
         """Create the blob directory and database if necessary, start all desired services"""
 
-        log.debug("Starting session.")
-
-        if self.node_id is None:
-            self.node_id = generate_id()
+        log.info("Starting session.")
 
         if self.wallet is None:
             from lbrynet.core.PTCWallet import PTCWallet
             self.wallet = PTCWallet(self.db_dir)
 
-        if self.peer_manager is None:
-            self.peer_manager = PeerManager()
-
         if self.use_upnp is True:
-            d = self._try_upnp()
-        else:
-            d = defer.succeed(True)
-
-        if self.peer_finder is None:
-            d.addCallback(lambda _: self._setup_dht())
-        else:
-            if self.hash_announcer is None and self.peer_port is not None:
-                log.warning("The server has no way to advertise its available blobs.")
-                self.hash_announcer = DummyHashAnnouncer()
-
-        d.addCallback(lambda _: self._setup_other_components())
-        return d
+            yield self._try_upnp()
+        elif not self.external_ip:
+            yield self._get_external_ip()
+
+        log.info("Starting DHT")
 
+        hosts = []
+        for host, port in self.known_dht_nodes:
+            h = yield reactor.resolve(host)
+            hosts.append((h, port))
+
+        yield self.dht_node.joinNetwork(tuple(hosts))
+        self.peer_finder.run_manage_loop()
+        self.hash_announcer.run_manage_loop()
+
```

Comment:
Can't do self.hash_announcer.run_manage_loop() without running self.blob_manager.setup() first since hash announcer depends on blob manager.

Reply:
How can we make this dependency more real? How can we get the daemon to error when the dependency is not  met?